### DISCUSSION TITLE
Implication and Circuit constraints

### DISCUSSION THREAD

**Discussion by opto5897** [google/or-tools#3924] [CP-SAT questions] [answered]
Hi guys! While studying constraints in OR Tools I found an example 
https://or.stackexchange.com/questions/10214/about-circuit-constraints-in-or-tools
which is very similar to my problem. There is already a very good answer. I get the logic, and the whole code is just a dozen lines, but as a newbie I cannot combine the answer and the OP code to fix it. So I cannot get a working example. Please, could someone post a complete code? Or maybe someone already has a gist. I think it will be useful for other beginners like me. Thank you.

**Comment by lperron** [comment]
this is too vague. I do not even understand what you want. What is the `OP code` ?

**Reply by opto5897** [reply]
I'm sorry. I mean that the question which I mentioned contains example code which does not work. Answer contains instructions how to fix the code. But I cannot follow these instructions.

**Reply by Mizux** [reply]
OP: Orignal Post so I guess this user want to merge the question snippet code with your solution...

**Reply by opto5897** [reply]
Please, need some hints. I attached the code from my attempt.

```python
from ortools.sat.python import cp_model

model = cp_model.CpModel()

bool_node_active = {
    0: model.NewBoolVar("Is node 0 active"),
    1: model.NewBoolVar("Is node 1 active"),
    2: model.NewBoolVar("Is node 2 active"),
}

bool_arcs = {}
for i in range(3):
    for j in range(3):
        bool_arcs[(i, j)] = model.NewBoolVar("")

arcs = []
for i in range(3):
    for j in range(3):
        if i == j:
            # Each node can be optional "(k, k, lit.Not())"
            arcs.append([i, j, bool_arcs[i, j].Not()])
        else:
            arcs.append([i, j, bool_arcs[i, j]])

for i in range(3):
    for j in range(3):
        model.AddImplication(bool_arcs[i, j], bool_node_active[i])
        model.AddImplication(bool_arcs[i, j], bool_node_active[j])

# The case where the resource is empty (no tasks). We need dummy node 0
empty = model.NewBoolVar('empty')
arcs.append([0, 0, empty])
for i in range(3):
    model.AddImplication(empty, bool_node_active[i].Not())
        
model.AddCircuit(arcs)

# Solve model
solver = cp_model.CpSolver()
status = solver.Solve(model)
print('Status:', status) 
# Status: 1
```

**Comment by lperron** [comment, marked as answer]
Here is a working model.

```python
from ortools.sat.python import cp_model

num_nodes = 3

model = cp_model.CpModel()
bool_node_active = [model.NewBoolVar(f'bool_node_active_{i}') for i in range(num_nodes)]

arcs = []
for i in range(num_nodes):
    # if node i is first.
    start_lit = model.NewBoolVar(f'start_{i}')
    arcs.append((0, i + 1, start_lit))

    # if node i is last.
    end_lit = model.NewBoolVar(f'end_{i}')
    arcs.append((i + 1, 0, end_lit))
    
    for j in range(num_nodes):
        if i == j:
            arcs.append((i + 1, i + 1, bool_node_active[i].Not()))
        else:
            literal = model.NewBoolVar(f'arc_{i}_to_{j}')
            arcs.append((i + 1, j + 1, literal))

# Manage the empty circuit
empty = model.NewBoolVar('circuit')
arcs.append((0, 0, empty))

for i in range(num_nodes):
    model.AddImplication(empty, bool_node_active[i].Not())

# Add the circuit constraint.
model.AddCircuit(arcs)

# Solve model
solver = cp_model.CpSolver()
status = solver.Solve(model)
print('Status:', status) 
```

**Reply by opto5897** [reply]
Hello Laurent,

Huge thanks! This example is a great learning for me.
Currently I read many discussion topics, and I want to say that your and Mizux's contribution to the community is invaluable. Thank you.

**Comment by lperron** [comment]
you're welcome. I have added the examples to the sat samples

https://github.com/google/or-tools/blob/main/ortools/sat/docs/scheduling.md#ranking-tasks-in-a-disjunctive-resource-with-a-circuit-constraint
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00
